### ISSUE TITLE
Problem with Prophet model regressor  

### ISSUE THREAD

**Issue by vjosapreniqi** [facebook/prophet#1134]
Hi,
I am working on predicting half-hourly UK electricity prices with prophet. I have two other time series: gas prices and Initial national demand out-turn. So, after merging all the data-sets together  based on datetime index, I used gas prices and initial national demand as regressors. When I use only gas as regressor, I get better prediction results. However, If I add load demand, the model gives worse results. I am a bit confused cuz both regressors are positively correlated and have impact on electricity prices.  Also, one important detail is that the load demand has very clear pattern compared to electricity and gas prices series. So, from the prediction graph, it seems like prophet model is following that pattern that is easier to follow, which in this case is load demand.  Anyhow, I might be wrong somewhere.
I will attach some of the code : 
#Prophet:
model_P = Prophet(weekly_seasonality=True,yearly_seasonality=False, daily_seasonality=True)
model_P.add_seasonality(name="monthly", period=30.5, fourier_order=5)
model_P.add_regressor('Gas')
model_P.add_regressor('INDO')
model_P.fit(pd.DataFrame({'ds': train.index, 'y':train, 'Gas': train_gas, 'INDO': train_indo}))

def populate_indo_val(dt):
    current_date=str(dt.date())
    return df[current_date]['INDO'][0]

def populate_gas_val(dt):
    current_date=str(dt.date())
    return df[current_date]['Gas'][0]

future_dates = model_P.make_future_dataframe(periods=test.shape[0],freq='30T')
future_dates['INDO']=future_dates['ds'].apply(populate_indo_val)
future_dates['Gas']=future_dates['ds'].apply(populate_gas_val)
results_P = model_P.predict(future_dates)
forecast_P = results_P.set_index('ds').yhat
`
I am Also attaching the forecasting model graphs, when only gas is used as regressor and, the second graph depicts the model where only initial load deamand is used for prediction prediction.
<img width="631" alt="Prophet graphs" src="https://user-images.githubusercontent.com/12625194/64344839-29f7a000-cfe7-11e9-93f0-f5dbab8e8817.png">

**Comment by bletham** [comment]
So from the figures, with Gas as a regressor the fit looks good, with INDO it looks poor. If you do both Gas and INDO at the same time does it look like the Gas only figure? That'd be my expectation but it sounds from your description like that might not be the case.

I'm not really sure what is causing the bad fit with INDO, is there any chance you can post the data for these time series so I can fit the models and dig a little deeper?

**Comment by vjosapreniqi** [comment]
Hi, 

Thank you very much

 for the feedback! I have tried to put both gas and INDO as external regressors but only small improvement was done, and the prediction graph looks similar to INDO only figure.
Below you have the time-series where INDO and MID are half-hourly recorded, while gas series is daily recorded.
[gas_data.xlsx](https://github.com/facebook/prophet/files/3743845/gas_data.xlsx)
[INDO-2017-2019.xlsx](https://github.com/facebook/prophet/files/3743846/INDO-2017-2019.xlsx)
[mid-2017-2019-datetime.xlsx](https://github.com/facebook/prophet/files/3743847/mid-2017-2019-datetime.xlsx)

**Comment by bletham** [comment]
Sorry for the delay in looking into this - 

I'm not able to produce the bad fit with INDO. I wonder if something may have gone wrong when merging the dataframes? This is my complete code starting from the xlsx files you provided:
```
import pandas as pd
from fbprophet import Prophet

df_data = pd.read_excel('mid-2017-2019-datetime.xlsx')
df_gas = pd.read_excel('gas_data.xlsx')
df_indo = pd.read_excel('INDO-2017-2019.xlsx')


df = pd.DataFrame({
    'ds': df_data['datetime'],
    'y': df_data['Price'],
})
df['ds'] = pd.to_datetime(df['ds'])
# Merge in INDO
df_indo['ds'] = pd.to_datetime(df_indo['datetime'])
df = pd.merge(df, df_indo[['ds', 'INDO']], on=['ds'], how='left')
# Merge in gas
df['date'] = pd.to_datetime(df['ds'].dt.date)
df_gas['date'] = pd.to_datetime(df_gas['Date'])
df_gas = df_gas[['date', 'Gas (£/MWh)']]
df_gas.columns = ['date', 'Gas']
df = pd.merge(df, df_gas[['date', 'Gas']], on=['date'], how='left')
df = df[['ds', 'y', 'INDO', 'Gas']]

# Restrict to 2019 as in the plots in the issue
df2 = df[df['ds'] >= pd.to_datetime('2019-01-01')]

# Fit a model with no regressors
m1 = Prophet(weekly_seasonality=True,yearly_seasonality=False, daily_seasonality=True)
m1.add_seasonality(name="monthly", period=30.5, fourier_order=5)
m1.fit(df2)
fcst1 = m1.predict()
m1.plot(fcst1)

# Gas regressor
m2 = Prophet(weekly_seasonality=True,yearly_seasonality=False, daily_seasonality=True)
m2.add_seasonality(name="monthly", period=30.5, fourier_order=5)
m2.add_regressor('Gas')
m2.fit(df2)
fcst2 = m2.predict()
m2.plot(fcst2)
m2.plot_components(fcst2)

# INDO regressor
m3 = Prophet(weekly_seasonality=True,yearly_seasonality=False, daily_seasonality=True)
m3.add_seasonality(name="monthly", period=30.5, fourier_order=5)
m3.add_regressor('INDO')
m3.fit(df2)
fcst3 = m3.predict()
m3.plot(fcst3)
m3.plot_components(fcst3)

# Both regressors
m4 = Prophet(weekly_seasonality=True,yearly_seasonality=False, daily_seasonality=True)
m4.add_seasonality(name="monthly", period=30.5, fourier_order=5)
m4.add_regressor('INDO')
m4.add_regressor('Gas')
m4.fit(df2)
fcst4 = m4.predict()
m4.plot(fcst4)
m4.plot_components(fcst4)
```

This is the plot I get for the INDO only fit (m3 above):
![INDO](https://user-images.githubusercontent.com/6339760/68348838-01a42480-00b0-11ea-957c-7f38083f08cf.png)

If you run my code does it work for you? Can you tell what was done differently in your run?

**Comment by bletham** [comment]
As a side note, this is a really nice example for extra regressors - there is a strong fitted effect for both, but especially for Gas.

**Comment by vjosapreniqi** [comment]
Thank you very much for your feedback @bletham! I am producing the same output as you for the training set. But when I try to predict values in the future, using INDO as an external regressor, I get the bad result as I have explained in the first comment.
More specifically: I have divided the dataset into the training set and testing set, I have hidden the test set from the model in order to evaluate the performance, and I have predicted the values (treating them as 'future data' ) for the same timeframe as in the test set. However, I have let the model know the future values of gas or/and INDO. 

To do so, I have used this code: 

def populate_indo_val(dt):
   current_date=str(dt.date())
   return df[current_date]['INDO'][0]

future_dates = model_P.make_future_dataframe(periods=test.shape[0],freq='30T')
future_dates['INDO']=future_dates['ds'].apply(populate_indo_val)
results_P = model_P.predict(future_dates)
forecast_P = results_P.set_index('ds').yhat    

I guess I am going wrong somewhere while trying to populate the 'future data frame' with INDO values (method populate_indo_val(dt)).

**Comment by bletham** [comment]
hm I don't see anything obviously wrong with the code, but if you provide code that I can run end-to-end (that is, starting from the xlsx file and containing the train/test splitting) I could look more closely and try and see what is happening.

(Sorry for the long delay that has likely made this response useless).

**Comment by vjosapreniqi** [comment]
Hi @bletham, sure I can provide you with the files. Lately, a have posted the entire project on Github so here are the links:
All files: https://github.com/vjosapreniqi/UK-electricity-price-prediction
 The prophet model file: https://github.com/vjosapreniqi/UK-electricity-price-prediction/blob/master/Prophet%20Models.ipynb
 Cheers :)

**Comment by bletham** [comment]
OK, the issue here was a bit subtle but I finally figured it out. Fortunately not a bug. It's coming from how INDO values are being set in the `future_dates` dataframe. If you run this that includes just INDO, you'll see that you get the bad predictions with the large bias:

```
model_P = Prophet(weekly_seasonality=True, daily_seasonality=True, yearly_seasonality=False, holidays=holidays)
model_P.add_seasonality(name="monthly", period=30.5, fourier_order=5)
model_P.add_regressor('INDO')
model_P.fit(pd.DataFrame({'ds': train.index, 'y':train,'INDO': train_indo}))


def populate_indo_val(dt):
   current_date=str(dt.date())
   return df[current_date]['INDO'][0]

future_dates = model_P.make_future_dataframe(periods=test.shape[0],freq='30T')
future_dates['INDO']=future_dates['ds'].apply(populate_indo_val)
results_P = model_P.predict(future_dates)
fig=model_P.plot(results_P)
```
The problem is with `populate_indo_val`. For Gas, this function is correct and assigns for each day the value of Gas for that day. For INDO, it doesn't work because in the training data (given to model_P.fit), INDO takes on different values throughout the day. But populate_indo_val is holding it constant throughout the day (at whatever the value was at the start of the day). This means the predictions are using different INDO values than what the model was fit with. In this particular case, INDO has pretty strong daily seasonality and is low at the start of the day, so we were making predictions with INDO values much lower than they should have been, which biased the whole forecast down. The fix is easy - you just need to use the half-hourly data for INDO in `future_dates` the same way it is being used in `model_P.fit`. This code gives a great fit:

```
model_P = Prophet(weekly_seasonality=True, daily_seasonality=True, yearly_seasonality=False, holidays=holidays)
model_P.add_seasonality(name="monthly", period=30.5, fourier_order=5)
model_P.add_regressor('INDO')
model_P.add_regressor('Gas')
model_P.fit(pd.DataFrame({'ds': train.index, 'y':train, 'INDO': train_indo, 'Gas': train_gas}))

def populate_gas_val(dt):
   current_date=str(dt.date())
   return df[current_date]['Gas'][0]

future_dates = model_P.make_future_dataframe(periods=test.shape[0],freq='30T')
df2 = pd.DataFrame({'ds': df.index, 'INDO': df['INDO']})
future_dates = pd.merge(future_dates, df2, on='ds', how='inner')
future_dates['Gas']=future_dates['ds'].apply(populate_gas_val)

results_P = model_P.predict(future_dates)
```

Let me know if that makes sense, and thanks again for your patience in getting this worked out.

**Comment by vjosapreniqi** [comment]
Thank you very much @bletham. I tried this and it works perfectly well now!
Thank you very much for your explanations!
Cheers.
